### PR TITLE
MYNN-EDMF update

### PR DESCRIPTION
**Pull request by joeolson42**
MYNN-EDMF update including MYNN-related modules

1) Updates to the MYNN-EDMF from RRFS_dev, which include updates made for RRFS "Base E" ([pdf](https://drive.google.com/file/d/103uaNc3tRkt5eS544PgLcoRbOkCp_ADY/view?usp=share_link) here)+ some addition mods below:
2) Including snow mixing. This required passing snow into the MYNN, modifying the negative mixing ratio checks, and adding a new tendency array for output.
3) Includes a patch for ensuring cloud fractions for all grid cells where qc > 1e-6 and qi > 1e-9. See [pdf](https://drive.google.com/file/d/1Y0FqCr5UfW6WuWGKSQG4QCLqOHK9z22j/view?usp=share_link).
4) Misc code cleaning: removing some test code, removing from if-check from within loops, cleaning up indentations and interfaces, etc.
5) Updates to the sgscloud_radpre module for use with MYNN. This removes a bug for supplying grid mean mixing ratios and also adda a section for use with SAS convection scheme and imports the Ferrier-Aligo flag for future coupling tests.
6) Included many changes to control the specified model precision.

Companion PRs:
https://github.com/NOAA-EMC/fv3atm/pull/630
https://github.com/ufs-community/ufs-weather-model/pull/1636

### CODE REVIEW

**Comment by yangfanglin** [comment]
Joe,  could you please add a link to the ppt about these updates you presented at the RRFS meeting ? Thx

**Comment by ericaligo-NOAA** [comment]
Joe, nice talk the other day. You said there was no noticeable differences with a winter case. I'm wondering if one would see a difference in the onset of snow at the surface when the snow is included in the mixing. Some years ago Brad Ferrier and I thought about mixing snow in the MYJ PBL. The NAM tended to have a delay in the onset of snow reaching the ground. Vertical cross sections showed it was dry in the lowest few hundred mbs, but RH was near 100% above that. We wondered if mixing snow would increase the RH sooner in the lowest levels.

**Comment by joeolson42** [comment]
Fanglin,

I have update the description to include links to some pdfs.

Eric,

I should clarify that I didn't see any differences in the overall cloud cover (this wasn't the ultimate fix for the "missing cloud fraction" problem) but there may be some other small impacts, like you mention. I'll definitely take a closer look. We will also be running more winter retros soon. Summer retros have been getting most of the CPU hrs.

-joe

**Comment by ericaligo-NOAA** [comment]
Thanks, Joe! It will be interesting to see if there is a difference in snow onset for the winter cases.

File: physics/module_bl_mynn.F90 [hunk 1]
```diff
@@ -301,6 +291,7 @@ MODULE module_bl_mynn
 !      &cns=3.5, alp1=0.23, alp2=0.3, alp3=3.0, alp4=10.0, alp5=0.2
 
   REAL, PARAMETER :: gpw=5./3., qcgmin=1.e-8, qkemin=1.e-12
+  REAL, PARAMETER :: tliq = 269. !all hydrometeors are liquid when T > tliq
```
Comment (dustinswales) [inline physics/module_bl_mynn.F90]: I recommend these module parameters be declared using the kind_phys precision.

File: physics/module_bl_mynn.F90 [hunk 2]
```diff
@@ -480,71 +444,67 @@ SUBROUTINE mynn_bl_driver(            &
...
+   REAL(kind=kind_phys), DIMENSION(:,:), INTENT(inout) ::         &
+         & edmf_a,edmf_w,edmf_qt,edmf_thl,edmf_ent,edmf_qc,       &
          & sub_thl3D,sub_sqv3D,det_thl3D,det_sqv3D
 
 !   REAL, DIMENSION(IMS:IME,KMS:KME)   :: &
 !         & edmf_a_dd,edmf_w_dd,edmf_qt_dd,edmf_thl_dd,edmf_ent_dd,edmf_qc_dd
 
-    REAL, DIMENSION(:), INTENT(inout) :: Pblh,rmol
+    REAL(kind=kind_phys), DIMENSION(:), INTENT(inout) :: Pblh
+    REAL, DIMENSION(:), INTENT(inout) :: rmol
```
Comment (dustinswales) [inline physics/module_bl_mynn.F90]: Use real(kind_phys)

File: physics/module_bl_mynn.F90 [hunk 3]
```diff
@@ -480,71 +444,67 @@ SUBROUTINE mynn_bl_driver(            &
...
+         &dqke1,diss_heat
 
-    REAL, DIMENSION(:,:), intent(out) :: Sh3D,Sm3D
+    REAL(kind=kind_phys), DIMENSION(:,:), intent(out) :: Sh3D,Sm3D
 
-    REAL, DIMENSION(:,:), INTENT(inout) ::      &
+    REAL(kind=kind_phys), DIMENSION(:,:), INTENT(inout) ::        &
          &qc_bl,qi_bl,cldfra_bl
-    REAL, DIMENSION(KTS:KTE) :: qc_bl1D,qi_bl1D,cldfra_bl1D,&
+    REAL, DIMENSION(KTS:KTE) :: qc_bl1D,qi_bl1D,cldfra_bl1D,      &
```
Comment (dustinswales) [inline physics/module_bl_mynn.F90]: Use kind_phys

File: physics/module_bl_mynn.F90 [hunk 4]
```diff
@@ -480,71 +444,67 @@ SUBROUTINE mynn_bl_driver(            &
...
+    REAL(kind=kind_phys), DIMENSION(:,:), INTENT(inout) :: el_pbl
 
-    REAL, DIMENSION(:,:), INTENT(out) ::          &
+    REAL(kind=kind_phys), DIMENSION(:,:), INTENT(out) ::          &
          &qWT,qSHEAR,qBUOY,qDISS,dqke
-    ! 3D budget arrays are not allocated when bl_mynn_tkebudget == .false.
+    ! 3D budget arrays are not allocated when tke_budget == 0
     ! 1D (local) budget arrays are used for passing between subroutines.
-    REAL, DIMENSION(kts:kte) :: qWT1,qSHEAR1,qBUOY1,qDISS1,dqke1,diss_heat
+    REAL, DIMENSION(kts:kte) :: qwt1,qshear1,qbuoy1,qdiss1,       &
```
Comment (dustinswales) [inline physics/module_bl_mynn.F90]: Use real(kind_phys)

File: physics/module_bl_mynn.F90 [hunk 5]
```diff
@@ -480,71 +444,67 @@ SUBROUTINE mynn_bl_driver(            &
...
-!    REAL,    DIMENSION( ims:ime, kms:kme, nchem ), INTENT(INOUT), optional :: chem3d
-!    REAL,    DIMENSION( ims:ime, kdvel, ndvel ), INTENT(IN), optional :: vdep
-    REAL,    DIMENSION(:, :, :), INTENT(INOUT) :: chem3d
-    REAL,    DIMENSION(:, :),   INTENT(IN) :: vdep
-    REAL,    DIMENSION(:),     INTENT(IN) :: frp,EMIS_ANT_NO
+    REAL(kind=kind_phys), DIMENSION(:,:,:), INTENT(INOUT) :: chem3d
+    REAL(kind=kind_phys), DIMENSION(:,:),   INTENT(IN)    :: vdep
+    REAL(kind=kind_phys), DIMENSION(:),     INTENT(IN)    :: frp,EMIS_ANT_NO
     !local
     REAL,    DIMENSION(kts:kte  ,nchem) :: chem1
```
Comment (dustinswales) [inline physics/module_bl_mynn.F90]: Use real(kind_phys) for all local declarations.

File: physics/module_bl_mynn.F90 [hunk 6]
```diff
@@ -711,7 +688,7 @@ SUBROUTINE mynn_bl_driver(            &
           ENDDO
        ENDDO
 
-       IF ( bl_mynn_tkebudget ) THEN
+       IF (tke_budget .eq. 1) THEN
```
Comment (dustinswales) [inline physics/module_bl_mynn.F90:699]: Why the change from a logical to integer? (It seems that tke_budget can only be 0/1)

File: physics/module_bl_mynn.F90 [hunk 7]
```diff
@@ -1607,47 +1483,44 @@ END SUBROUTINE mynn_bl_driver
...
 !
 !-------------------------------------------------------------------
     
     INTEGER, INTENT(IN)   :: kts,kte
-    INTEGER, INTENT(IN)   :: bl_mynn_mixlength,bl_mynn_edmf
+    INTEGER, INTENT(IN)   :: bl_mynn_mixlength
     LOGICAL, INTENT(IN)   :: INITIALIZE_QKE
 !    REAL, INTENT(IN)   :: ust, rmo, pmz, phh, flt, flq
-    REAL, INTENT(IN)   :: ust, rmo, Psig_bl, dx
+    REAL, INTENT(IN)   :: rmo, Psig_bl, xland
```
Comment (dustinswales) [inline physics/module_bl_mynn.F90]: Use real(kind_phys)

File: physics/module_bl_mynn.F90 [hunk 8]
```diff
@@ -2733,69 +2588,70 @@ END SUBROUTINE boulac_length
...
 
     REAL ::  a2fac, duz, ri !JOE-Canuto/Kitamura mod
 
-    REAL:: auh,aum,adh,adm,aeh,aem,Req,Rsl,Rsl2,&
-           gmelq,sm20,sh20,sm25max,sh25max,sm25min,sh25min,&
+    REAL:: auh,aum,adh,adm,aeh,aem,Req,Rsl,Rsl2,                &
+           gmelq,sm20,sh20,sm25max,sh25max,sm25min,sh25min,     &
            sm_pbl,sh_pbl,zi2,wt,slht,wtpr
 
     DOUBLE PRECISION  q2sq, t2sq, r2sq, c2sq, elsq, gmel, ghel
```
Comment (dustinswales) [inline physics/module_bl_mynn.F90:2664]: Maybe overkill for now, but it would be ideal at some point if DOUBLE PRECISION was replaced with the parameterized type double, kind_dbl_prec. This is there to provide control when specific variables within a scheme need to be defined with a specific precision [e.g. random number generation in radiation is another instance ](https://github.com/ufs-community/ccpp-physics/blob/ufs/dev/physics/rrtmgp_sw_main.F90#L225).

File: physics/module_bl_mynn.F90 [hunk 9]
```diff
@@ -4352,7 +4184,8 @@ SUBROUTINE mynn_tendencies(kts,kte,i,    &
     d(kte)=u(kte)
 
 !    CALL tridiag(kte,a,b,c,d)
-    CALL tridiag3(kte,a,b,c,d,x)
+    CALL tridiag2(kte,a,b,c,d,x)
+!    CALL tridiag3(kte,a,b,c,d,x)
```
Comment (dustinswales) [inline physics/module_bl_mynn.F90:4199]: These and subsequent commented lines around the calls to the tridiag routines are probably not necessary.

File: physics/sgscloud_radpre.F90 [hunk 10]
```diff
@@ -35,16 +35,19 @@ module sgscloud_radpre
 !!
 !>\section sgscloud_radpre_mod  SGS Cloud Scheme Pre General Algorithm
       subroutine sgscloud_radpre_run(    &
```
Comment (dustinswales) [inline physics/sgscloud_radpre.F90:37]: Not that this matters for this PR, but RRTMGP doesn't use sgscloud_radpre_run, so these change will ultimately needed to be ported to [GP's cloud coupling](https://github.com/ufs-community/ccpp-physics/blob/ufs/dev/physics/GFS_rrtmgp_cloud_mp.F90).

Looking at this now it seems that they diverged sometime ago, so nbd.

**Review by dustinswales — COMMENTED** [review]
I don't have any major concerns, just the mismatch of the working precision in the scheme, which should get resolved. I know it's an invasive step, so I'm fine with it being part of a follow up PR.

File: physics/module_bl_mynn.F90 (hunk 1 above)
Comment (joeolson42) [reply physics/module_bl_mynn.F90]: I will be committing some changes soon. This has been changed as requested.

File: physics/module_bl_mynn.F90 (hunk 2 above)
Comment (joeolson42) [reply physics/module_bl_mynn.F90]: done

File: physics/module_bl_mynn.F90 (hunk 3 above)
Comment (joeolson42) [reply physics/module_bl_mynn.F90]: done

File: physics/module_bl_mynn.F90 (hunk 4 above)
Comment (joeolson42) [reply physics/module_bl_mynn.F90]: done

File: physics/module_bl_mynn.F90 (hunk 5 above)
Comment (joeolson42) [reply physics/module_bl_mynn.F90]: done

File: physics/module_bl_mynn.F90 (hunk 6 above)
Comment (joeolson42) [reply physics/module_bl_mynn.F90:699]: Yes, this one is annoying. I had to change it to an integer because, in WRF, this flag is used as a package declaration (similar to how it's used in GFS_typedefs.F90, where 5 3D arrays are initialized when this is set to 1. However, WRF preprocessing can not handle logical flags, so I had to change it back to integer. Part of the many compromises for "universalization".

File: physics/module_bl_mynn.F90 (hunk 7 above)
Comment (joeolson42) [reply physics/module_bl_mynn.F90]: done, will be committed after test case is run.

File: physics/module_bl_mynn.F90 (hunk 8 above)
Comment (joeolson42) [reply physics/module_bl_mynn.F90:2664]: I'll shelve this one for the time being.

File: physics/module_bl_mynn.F90 (hunk 9 above)
Comment (joeolson42) [reply physics/module_bl_mynn.F90:4199]: I still want to test the impact of these different solvers more systematically first. I haven't found much difference, but supposedly, tridiag2 is 11% faster than tridiag. As soon as I test this more thoroughly, I'll clean it up.

File: physics/sgscloud_radpre.F90 (hunk 10 above)
Comment (joeolson42) [reply physics/sgscloud_radpre.F90:37]: Yes, I'd like this to be cleaned up. Done once for all configurations. Unfortunately, I need a lot of time to soak up what we have and know best what to propose in it's place.

File: physics/module_bl_mynn.F90 (hunk 6 above)
Comment (dustinswales) [reply physics/module_bl_mynn.F90:699]: Gotcha. I wasn't aware of the constraints from the WRF pp.

File: physics/mynnedmf_wrapper.F90 [hunk 11]
```diff
@@ -25,25 +25,25 @@ subroutine mynnedmf_wrapper_init (          &
...
-        real(kind=kind_phys),intent(in)  :: con_cp
-        real(kind=kind_phys),intent(in)  :: con_cpv
-        real(kind=kind_phys),intent(in)  :: con_rcp
-        real(kind=kind_phys),intent(in)  :: con_p608
-        real(kind=kind_phys),intent(in)  :: con_cliq
-        real(kind=kind_phys),intent(in)  :: con_cice
-        real(kind=kind_phys),intent(in)  :: con_karman
-        real(kind=kind_phys),intent(in)  :: con_t0c
+        logical,        intent(in)  :: do_mynnedmf
+        logical,        intent(in)  :: lheatstrg
```
Comment (grantfirl) [inline physics/mynnedmf_wrapper.F90]: Since the check for `lheatstrg` has been removed and it isn't used elsewhere, we can remove it from the argument list and metadata.

File: physics/mynnedmf_wrapper.F90 [hunk 12]
```diff
@@ -140,16 +135,18 @@ SUBROUTINE mynnedmf_wrapper_run(        &
      &  nupdraft,maxMF,ktop_plume,      &
      &  dudt, dvdt, dtdt,                                  &
      &  dqdt_water_vapor,            dqdt_liquid_cloud,    & ! <=== ntqv, ntcw
-     &  dqdt_ice_cloud,              dqdt_ozone,           & ! <=== ntiw, ntoz
+     &  dqdt_ice_cloud,              dqdt_snow_cloud,      & ! <=== ntiw, ntsw
```
Comment (grantfirl) [inline physics/mynnedmf_wrapper.F90]: We should have no say about what variables are called locally, but I'm wondering why this is `dqdt_snow_cloud` instead of just `dqdt_snow`. It's referring to the tendency of the snow tracer which presumably is grid mean and sediments out of the cloud.

File: physics/mynnedmf_wrapper.F90 [hunk 13]
```diff
@@ -588,21 +540,38 @@ SUBROUTINE mynnedmf_wrapper_run(        &
...
+             rho(i,k)=prsl(i,k)/(r_d*t3d(i,k)*(1.+p608*max(sqv(i,k),1e-8)))
              w(i,k) = -omega(i,k)/(rho(i,k)*grav)
+          enddo
+       enddo
+
+  ! Check incoming moist species to ensure non-negative values
+  ! First, create height difference (dz)
+      do k=1,levs
+         do i=1,im
+            dz(i,k)=(phii(i,k+1) - phii(i,k))*g_inv
```
Comment (grantfirl) [inline physics/mynnedmf_wrapper.F90:555]: It's amazing to me that `dz` and `delp` don't already exist as CCPP variables to pass in. It looks like RRTMGP calculates/saves them, but other schemes, if they do calculate them (and I'm sure that some do), they're not sharing them.

File: physics/sgscloud_radpre.F90 [hunk 14]
```diff
@@ -35,16 +35,19 @@ module sgscloud_radpre
...
            flag_init,flag_restart,       &
            con_g, con_pi, eps, epsm1,    &
            r_v, cpv, rcp,                &
            xlv, xlf, cp,                 &
            do_mynnedmf,                  &
            qc, qi, qv, T3D, P3D, exner,  &
            qr, qs, qg,                   &
-           qci_conv,ud_mf,               &
+           qci_conv,qlc,qli,ud_mf,       &
+!           qci_conv_timeave,             &
```
Comment (grantfirl) [inline physics/sgscloud_radpre.F90]: Any reason to leave these commented lines?

File: physics/sgscloud_radpre.F90 [hunk 15]
```diff
@@ -67,18 +71,20 @@ subroutine sgscloud_radpre_run(    &
...
       logical,             intent(in)  :: flag_init, flag_restart, do_mynnedmf
 
       real(kind=kind_phys), dimension(:,:), intent(inout) :: qc, qi
       real(kind=kind_phys), dimension(:,:), intent(inout) :: qr, qs, qg
-      ! qci_conv only allocated if GF is used
+      ! note: qci_conv only allocated if GF is used
       real(kind=kind_phys), dimension(:,:), intent(inout) :: qci_conv
+      real(kind=kind_phys), dimension(:,:), intent(inout) :: qlc, qli !for SAS
       real(kind=kind_phys), dimension(:,:), intent(in)    :: ud_mf
+      !real(kind=kind_phys), dimension(:,:), intent(in)    :: ud_mf_timeave, qci_conv_timeave
```
Comment (grantfirl) [inline physics/sgscloud_radpre.F90]: ditto

File: physics/mynnedmf_wrapper.F90 (hunk 13 above)
Comment (ChunxiZhang-NOAA) [reply physics/mynnedmf_wrapper.F90:555]: Typically, each physics scheme calculate dz between full levels, and between half levels, dp between full levels when the scheme needs them. Keep in mind that geopotential height is updated in CCPP.

Comment (grantfirl) [reply physics/mynnedmf_wrapper.F90:555]: Yep, all good points, and good reason to recalculate as necessary in case the geopotential update ever changes order within the suite.

File: physics/module_bl_mynn.F90 [hunk 16]
```diff
@@ -736,51 +737,14 @@ SUBROUTINE mynn_bl_driver(            &
                 sqc(k)=sqc3D(i,k) !/(1.+qv(i,k))
                 sqv(k)=sqv3D(i,k) !/(1.+qv(i,k))
                 thetav(k)=th(i,k)*(1.+0.608*sqv(k))
```
Comment (grantfirl) [inline physics/module_bl_mynn.F90]: opportunity to use p608 rather than redefined constant

File: physics/module_bl_mynn.F90 [hunk 17]
```diff
@@ -885,654 +847,582 @@ SUBROUTINE mynn_bl_driver(            &
...
+          endif
+          !keep snow out for now - increases ceiling bias                                                                                                
+          sqw(k)= sqv(k)+sqc(k)+sqi(k)!+sqs(k)
+          thl(k)= th1(k) - xlvcp/ex1(k)*sqc(k) &
+               &         - xlscp/ex1(k)*(sqi(k)+sqs(k))
+          !Use form from Tripoli and Cotton (1981) with their
+          !suggested min temperature to improve accuracy.
+          !thl(k)=th(i,k)*(1.- xlvcp/MAX(tk1(k),TKmin)*sqc(k) &
+          !    &               - xlscp/MAX(tk1(k),TKmin)*sqi(k))
+          thetav(k)=th1(k)*(1.+0.608*sqv(k))
```
Comment (grantfirl) [inline physics/module_bl_mynn.F90]: ditto

File: physics/module_bl_mynn.F90 [hunk 18]
```diff
@@ -480,146 +444,166 @@ SUBROUTINE mynn_bl_driver(            &
...
-    REAL, DIMENSION(:), INTENT(OUT) ::                &
-         &maxmf
+    real(kind_phys), DIMENSION(:), INTENT(out) ::  maxmf
 
-    REAL, DIMENSION(:,:), INTENT(inout) ::      &
-         &el_pbl
+    real(kind_phys), DIMENSION(:,:), INTENT(inout) :: el_pbl
 
-    REAL, DIMENSION(:,:), INTENT(out) ::          &
+    real(kind_phys), DIMENSION(:,:), INTENT(out) ::          &
```
Comment (grantfirl) [inline physics/module_bl_mynn.F90]: All of these are only conditionally given values (`if tke_budget`). Since they're intent(out), they will be filled with garbage (rather than 0 or something else) when `tke_budget == .false.`. It probably doesn't matter since they're diagnostic, but it is one of the CCPP rules that intent(out) variables are initialized. Even though they are intent(inout) in the wrapper, they will be overwritten by garbage in here when `tke_budget == .false.`.

File: physics/module_bl_mynn.F90 [hunk 19]
```diff
@@ -2215,20 +2085,20 @@ SUBROUTINE  mym_length (                     &
...
-        alp3 = 1.5
+        alp3 = 2.0
         alp4 = 5.0
         alp5 = alp2 !like alp2, but for free atmosphere
         alp6 = 50.0 !used for MF mixing length
 
         ! Impose limits on the height integration for elt and the transition layer depth
         !zi2=MAX(zi,minzi)
-        zi2=MAX(zi,    200.)
+        zi2=MAX(zi,    300.)
```
Comment (grantfirl) [inline physics/module_bl_mynn.F90:2097]: ditto for these max/mins

File: physics/module_bl_mynn.F90 [hunk 20]
```diff
@@ -3003,10 +2865,16 @@ SUBROUTINE  mym_turbulence (                                &
        !IF ( sm(k) > sm25max ) sm(k) = sm25max
        !IF ( sm(k) < sm25min ) sm(k) = sm25min
        !sm(k) = Prnum*sh(k)
-       slht  = zi*0.1
-       wtpr  = min( max( (slht - zw(k))/slht, 0.0), 1.0) ! 1 at z=0, 0 above sfc layer
-       Prlim = 1.0*wtpr + (1.0 - wtpr)*Prlimit
-       sm(k) = MIN(sm(k), Prlimit*Sh(k))
+
+       !surface layer PR
+       !slht  = zi*0.1
```
Comment (grantfirl) [inline physics/module_bl_mynn.F90:2870]: Consider whether commenting out or deleting is more appropriate here. Git history can always save stuff like this if you want to make the code "prettier".

File: physics/module_bl_mynn.F90 [hunk 21]
```diff
@@ -3966,101 +3825,89 @@ SUBROUTINE  mym_condensation (kts,kte,  &
...
+           !cldfra_bl1D(K) = max(0., min(1., 0.5+0.36*atan(1.2*(q1(k)+0.4))))
+           !Best compromise: Improves marine stratus without adding much cold bias.
+           cldfra_bl1D(k) = max(0., min(1., 0.5+0.36*atan(1.8*(q1(k)+0.2))))
 
-           !CLOUD WATER AND ICE
+           ! Specify hydrometeors
+           ! JAYMES- this option added 8 May 2015
+           ! The cloud water formulations are taken from CB02, Eq. 8.
            IF (q1k < 0.) THEN        !unsaturated
-#ifdef SINGLE_PREC
```
Comment (grantfirl) [inline physics/module_bl_mynn.F90:4002]: Don't we need this `#ifdef SINGLE_PREC`?

File: physics/module_bl_mynn.F90 [hunk 22]
```diff
@@ -6193,25 +6080,41 @@ SUBROUTINE DMP_mf(                        &
...
+           &     + exc_heat
 !was       UPTHL(1,I)= UPTHV(1,I)/(1.+svp1*UPQT(1,I))  !assume no saturated parcel at surface
        UPTHL(1,I)=(THL(KTS)*DZ(KTS+1)+THL(KTS+1)*DZ(KTS))/(DZ(KTS)+DZ(KTS+1)) &
-           &     +exc_fac*UPW(1,I)*sigmaTH/sigmaW
+           &     + exc_heat
+
+       !calculate exc_moist by use of surface fluxes
+       exc_moist=exc_fac*UPW(1,I)*sigmaQT/sigmaW
+       !calculate exc_moist by conserving rh:
+!       tk_int  =(tk(kts)*dz(kts+1)+tk(kts+1)*dz(kts))/(dz(kts+1)+dz(kts))
```
Comment (grantfirl) [inline physics/module_bl_mynn.F90:6105]: Same comment about allowing git history to keep this stuff if you ever need it again rather than commenting. I guess it depends on how permanent you think this is...

**Review by grantfirl — APPROVED** [review]
Only minor comments from me. You definitely want to check if the `#ifdef SINGLE_PREC` is still needed in order to not break the recent single precision changes by NRL and Sam.

File: physics/mynnedmf_wrapper.F90 (hunk 12 above)
Comment (joeolson42) [reply physics/mynnedmf_wrapper.F90]: Yes, I just have been in a few discussions recently about "what is a cloud?" If snow settles beneath the "snow cloud", when does it become a cloud in the k-1 layer? When do non-zero cloud fractions become appropriate? From a modeling point of view, it really depends on the cloud fraction parameterization, but which one is correct?

Well, I prefer shorter names, so I don't mind removing "cloud" from the name.

File: physics/mynnedmf_wrapper.F90 (hunk 13 above)
Comment (joeolson42) [reply physics/mynnedmf_wrapper.F90:555]: I think this is just a comment, but yes, perhaps there is a small case for some redundancy.

File: physics/sgscloud_radpre.F90 (hunk 14 above)
Comment (joeolson42) [reply physics/sgscloud_radpre.F90]: I can remove them. They are really for an anticipated follow-up commit for the GF scheme. But they can be added at that point.

File: physics/module_bl_mynn.F90 (hunk 16 above)
Comment (joeolson42) [reply physics/module_bl_mynn.F90]: indeed.

File: physics/module_bl_mynn.F90 (hunk 18 above)
Comment (joeolson42) [reply physics/module_bl_mynn.F90]: I think I'll change them to inout, since in a more free world, other schemes may be allowed to modify the TKE budget.

File: physics/module_bl_mynn.F90 (hunk 19 above)
Comment (joeolson42) [reply physics/module_bl_mynn.F90:2097]: I don't follow the ditto. Maybe the sequence of the comments I'm reading is different from the order at which you made them?

File: physics/module_bl_mynn.F90 (hunk 21 above)
Comment (joeolson42) [reply physics/module_bl_mynn.F90:4002]: I'm going to guess this came from NRL (or Sam). I'm not sure of its origins. Nor did I intend to delete it. I'll get it back in just in case.

File: physics/module_bl_mynn.F90 (hunk 20 above)
Comment (joeolson42) [reply physics/module_bl_mynn.F90:2870]: Yeah, some things I need to keep until I think about it more. I really only want to delete things I know I want out permanently.

File: physics/mynnedmf_wrapper.F90 (hunk 11 above)
Comment (joeolson42) [reply physics/mynnedmf_wrapper.F90]: Good catch. This lheatstrg needs to go.

File: physics/module_bl_mynn.F90 (hunk 19 above)
Comment (grantfirl) [reply physics/module_bl_mynn.F90:2097]: I think that I am referring to the opportunity to define min/max variables/parameters rather than hard-coded values. It's completely up to you, but I typically point this out in reviews when hard-coded values change in PRs because it can signal that the hard-coded values are really tuning knobs deserving of making them variables/parameters. If this is a one time change, then, obviously, there is no need.

**Comment by grantfirl** [comment]
@joeolson42 I didn't catch this earlier, but please update all of your PR branches to the top of their respective develop branches: ufs/dev for ccpp/physics, develop for fv3atm, and develop for ufs-weather-model.
